File: packages/baseline-py/src/baseline_py/init/repository_root.py
"""Find the repository a project lives in."""

from pathlib import Path


def repository_root(project_root: Path) -> Path:
    """Return the nearest ancestor holding ``.git``, or the project itself.

    A project is often nested inside a repository about something else, and
    GitHub reads workflows only from that repository's root.
    """
    for directory in (project_root, *project_root.parents):
        if (directory / ".git").exists():
            return directory
    return project_root

File: packages/create-baseline/bin/python/repositoryRoot.mjs
import { access } from 'node:fs/promises'
import { dirname, resolve } from 'node:path'

/**
 * The nearest ancestor holding `.git`, or the project itself.
 *
 * A Python project nested inside a repository keeps its workflow at that
 * repository's root, because GitHub reads workflows nowhere else; the check
 * has to look where the workflow can actually be.
 */
export async function repositoryRoot(root) {
  let dir = root
  for (;;) {
    try {
      await access(resolve(dir, '.git'))
      return dir
    } catch {
      /* keep climbing */
    }
    const parent = dirname(dir)
    if (parent === dir) return root
    dir = parent
  }
}
